File: lab2-algebra-templates/zad2-java/src/main/java/org/example/GaloisField.java
package org.example;

import lombok.Getter;

public class GaloisField {
    public GaloisField(int charact, int value) {
        if (!NumberTools.isPrime(charact)) {
            throw new IllegalArgumentException("character isn't prime");
        }
        if (value < 0 || value >= charact) {
            throw new IllegalArgumentException("value < 0 or value >= charact");
        }
        this.charact = charact;
        this.value = value;
    }

    public GaloisField(final GaloisField rhs) {
        this.charact = rhs.charact;
        this.value = rhs.value;
    }

    public GaloisField negative(int charact, int value) {
        if (value <= 0) {
            return new GaloisField(charact, (charact - 1 + value) % charact);
        } else {
            return new GaloisField(charact, value);
        }
    }

    public GaloisField inverse() {
        if (NumberTools.gcd(this.value, this.charact).orElse(0) != 1) {
            throw new IllegalArgumentException("Value and character aren't coprime");
        }
        int inverted = NumberTools.extEuclid(this.value, this.charact);
        return negative(this.charact, inverted);
    }

    public GaloisField add(final GaloisField rhs) {
        if (this.charact != rhs.charact) {
            throw new IllegalArgumentException("characters doesn't match");
        }
        return new GaloisField(this.charact, (this.value + rhs.value + this.charact) % this.charact);
    }

    public GaloisField substract(final GaloisField rhs) {
        if (this.charact != rhs.charact) {
            throw new IllegalArgumentException("characters doesn't match");
        }
        return new GaloisField(this.charact, (this.value - rhs.value + this.charact) % this.charact);
    }

    public GaloisField multiply(final GaloisField rhs) {
        if (this.charact != rhs.charact) {
            throw new IllegalArgumentException("characters doesn't match");
        }
        return new GaloisField(this.charact, (int) ((long) this.value * rhs.value) % this.charact);
    }

    public GaloisField divide(final GaloisField rhs) {
        if (this.charact != rhs.charact) {
            throw new IllegalArgumentException("characters doesn't match");
        }
        var opposite = new GaloisField(rhs).inverse();
        return new GaloisField(this.charact, (int) ((long) this.value * opposite.value) % this.charact);
    }

    public boolean equals(final GaloisField lhs, final GaloisField rhs) {
        return lhs.charact == rhs.charact && lhs.value == rhs.value;
    }

    public boolean notEquals(final GaloisField lhs, final GaloisField rhs) {
        return !equals(lhs, rhs);
    }

    public boolean greater(final GaloisField lhs, final GaloisField rhs) {
        return lhs.charact == rhs.charact && lhs.value > rhs.value;
    }

    public boolean less(final GaloisField lhs, final GaloisField rhs) {
        return lhs.charact == rhs.charact && lhs.value < rhs.value;
    }

    public boolean greaterEqual(final GaloisField lhs, final GaloisField rhs) {
        return !less(lhs, rhs);
    }

    public boolean lessEqual(final GaloisField lhs, final GaloisField rhs) {
        return !greater(lhs, rhs);
    }

    @Override
    public String toString() {
        return "characteristic: " + this.charact + "; value: " + this.value;
    }

    @Getter
    private final int charact;

    private final int value;
}

File: lab2-algebra-templates/zad3-rust/src/galois_field.rs
use std::cmp::Ordering;
use std::fmt::{Display, Formatter};
use std::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Sub, SubAssign};
use crate::number_tools::{ext_euclid, gcd, is_prime};

pub struct GaloisField<const C: u32> {
    value: u32,
}

impl<const CHARACT: u32> GaloisField<CHARACT> {
    pub fn new(value: u32) -> Option<Self> {
        if !is_prime(CHARACT) {
            panic!("Character isn't prime");
        }

        return if value >= CHARACT {
            None
        } else {
            Some(GaloisField { value })
        };
    }

    pub fn characteristic(&self) -> u32 { CHARACT }

    pub fn negative(value: i64) -> u32 {
        return if value > 0 {
            value as u32
        } else {
            ((CHARACT as i64 - value - 1) % CHARACT as i64) as u32
        }
    }

    pub fn inverse(self) -> Option<Self> {
        if gcd(self.value, CHARACT).unwrap_or(0) != 1 {
            panic!("Value and character aren't coprime");
        }
        let inverted = ext_euclid(self.value, CHARACT);
        Self::new(Self::negative(inverted as i64))
    }
}

impl<const CHARACT: u32> From<GaloisField<CHARACT>> for u32 {
    fn from(rhs: GaloisField<CHARACT>) -> Self {
        rhs.value
    }
}

impl<const CHARACT: u32> Display for GaloisField<CHARACT> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "value: {}; characteristic: {}", self.value, CHARACT)
    }
}

impl<const CHARACT: u32> Add for GaloisField<CHARACT> {
    type Output = GaloisField<CHARACT>;

    fn add(self, rhs: Self) -> Self::Output {
        let temp = self.value as i64 + rhs.value as i64;
        Self::new((temp % CHARACT as i64) as u32).unwrap()
    }
}

impl<const CHARACT: u32> Sub for GaloisField<CHARACT> {
    type Output = GaloisField<CHARACT>;

    fn sub(self, rhs: Self) -> Self::Output {
        let temp = self.value as u64 + Self::negative(rhs.value as i64) as u64;
        Self::new((temp % CHARACT as u64) as u32).unwrap()
    }
}

impl<const CHARACT: u32> Mul for GaloisField<CHARACT> {
    type Output = GaloisField<CHARACT>;

    fn mul(self, rhs: Self) -> Self::Output {
        let temp = self.value as u64 * rhs.value as u64;
        Self::new((temp % CHARACT as u64) as u32).unwrap()
    }
}

impl<const CHARACT: u32> Div for GaloisField<CHARACT> {
    type Output = GaloisField<CHARACT>;

    fn div(self, rhs: Self) -> Self::Output {
        let rhs_inv = rhs.inverse().unwrap();
        let temp = self.value as u64 * rhs_inv.value as u64;
        Self::new((temp % CHARACT as u64) as u32).unwrap()
    }
}

impl<const CHARACT: u32> AddAssign for GaloisField<CHARACT> {
    fn add_assign(&mut self, rhs: Self) {
        self.value = ((self.value as u64 + rhs.value as u64) % CHARACT as u64) as u32
    }
}

impl<const CHARACT: u32> SubAssign for GaloisField<CHARACT> {
    fn sub_assign(&mut self, rhs: Self) {
        self.value = ((self.value as u64 + Self::negative(rhs.value as i64) as u64) % CHARACT as u64) as u32
    }
}

impl<const CHARACT: u32> MulAssign for GaloisField<CHARACT> {
    fn mul_assign(&mut self, rhs: Self) {
        self.value = ((self.value as u64 * rhs.value as u64) % CHARACT as u64) as u32
    }
}

impl<const CHARACT: u32> DivAssign for GaloisField<CHARACT> {
    fn div_assign(&mut self, rhs: Self) {
        let rhs_inv = rhs.inverse().unwrap();
        self.value = ((self.value as u64 * rhs_inv.value as u64) % CHARACT as u64) as u32
    }
}

impl<const CHARACT: u32> PartialEq for GaloisField<CHARACT> {
    fn eq(&self, other: &Self) -> bool {
        self.value == other.value
    }
}

impl<const CHARACT: u32> PartialOrd for GaloisField<CHARACT> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.value.partial_cmp(&other.value)
    }
}
